pageserver/src/tenant/mgr.rs

```diff
@@ -130,6 +131,18 @@ pub(crate) enum TenantsMapRemoveResult {
     InProgress(utils::completion::Barrier),
 }
 
+/// When resolving a TenantId to a shard, we may be looking for the 0th
+/// shard, or we might be looking for whichever shard holds a particular page.
+pub(crate) enum ShardSelector {
+    /// Only return the 0th shard, if it is present.  If a non-0th shard is present,
+    /// ignore it.
+    Zero,
+    /// Pick the first shard we find for the TenantId
```

Comment:
May be better question - how it is possible to have shard 1 whit having shard 0?

Reply:
This isn't for getting the 0th shard: "first" in this context means the shard that sorts first among the shards attached to this pageserver.  So if the pageserver has shards 0, 1, 2 it'll be shard 0, but if the pageservfer has shards 7,8 attached it'll be shard 7.

This mode is the "I don't know what shard I want" callers, which is currently the case for the initial timeline lookup on page service connections.

Reply:
Consider a tenant with 8 shards and 4 pageservers (never mind why, it can happen).

Some pageserver might hold shards 0 and 3.  It will receive connections from postgres related to both shard 0 and shard 3, but postgres doesn't say at the start of a connection which shard it relates to, so at the start of the connection the compute is the one saying "I don't know what shard I want" (because the protocol doesn't allow it to express that).

I experimented with changing the `pagestream ...` part of the connection handling to include a TenantShardId instead of just a TenantId (it's pretty easy), but I think we will later want to be able to multiplex multiple shards' work down one connection.  For example, we might choose to have a lots of shards to smooth out data balance, beyond the number of pageservers we have.  In that case postgres might in future share the same connection for shards that happen to be on the same pageserver, so we still wouldn't want to tightly bind one connection to one shard on the server side.